src/hooks/useBasicTokens.ts

```diff
@@ -0,0 +1,172 @@
+
+import { BasicPoolsContext } from "../providers/BasicPoolsProvider"
+import { Contract } from "ethcall"
+import ERC20_ABI from "../constants/abis/erc20.json"
+import { Erc20 } from "./../../types/ethers-contracts/Erc20.d"
+import { GaugeContext } from "../providers/GaugeProvider"
+import { MinichefContext } from "../providers/MinichefProvider"
+import { areGaugesActive } from "../utils/gauges"
+import { useActiveWeb3React } from "../hooks"
+import { useContext } from "react"
+import { useMulticallProvider } from "./useMulticallProvider"
+
+export type BasicToken = {
+  address: string
+  name: string
+  symbol: string
+  decimals: number
+  isLPToken: boolean
+  isSynthetic: boolean
+  typeAsset: PoolTypes
+}
+export type BasicTokens = Partial<{ [address: string]: BasicToken }> | null
+
+export const useBasicTokens = (): UseQueryResult<BasicTokens> => {
+  const { chainId } = useActiveWeb3React()
+  const basicPools = useContext(BasicPoolsContext)
+  const minichefData = useContext(MinichefContext)
+  const { gauges } = useContext(GaugeContext)
+  const { data: ethCallProvider } = useMulticallProvider()
+  const gaugesAreActive = areGaugesActive(chainId)
+  const lpTokens = new Set()
+  const tokenType: Partial<{ [tokenAddress: string]: PoolTypes }> = {}
+
+  return useQuery(["tokens"], async () => {
+    if (!chainId || !basicPools || !ethCallProvider) {
+      return null
+    }
+    const targetTokenAddresses = new Set(
+      Object.values(basicPools)
+        .map((pool) => {
+          lpTokens.add(pool.lpToken)
+          const tokensInPool = [
+            ...pool.tokens,
+            ...(pool.underlyingTokens || []),
+            pool.lpToken,
+          ]
+          Object.assign(
+            tokenType,
+            ...tokensInPool.map((address) => ({
+              [address]: pool.typeOfAsset,
+            })),
+          ) as Record<string, PoolTypes>
+          return tokensInPool
+        })
+        .flat(),
+    )
+    if (minichefData) {
+      // add minichef reward tokens
+      minichefData.allRewardTokens.forEach((address) => {
+        targetTokenAddresses.add(address)
+      })
+    }
+    if (gauges) {
+      // add gauge tokens
+      Object.values(gauges).forEach((gauge) => {
+        gauge.rewards.forEach(({ tokenAddress }) => {
+          targetTokenAddresses.add(tokenAddress)
+        })
+      })
+    }
+    if (SDL_WETH_SUSHI_LP_CONTRACT_ADDRESSES[chainId] && gaugesAreActive) {
+      // add sushi token
+      targetTokenAddresses.add(
+        SDL_WETH_SUSHI_LP_CONTRACT_ADDRESSES[chainId].toLowerCase(),
+      )
+    }
+    if (VOTING_ESCROW_CONTRACT_ADDRESS[chainId] && gaugesAreActive) {
+      // add voting escrow token
+      targetTokenAddresses.add(
+        VOTING_ESCROW_CONTRACT_ADDRESS[chainId].toLowerCase(),
+      )
+    }
+    const tokenInfos = await getTokenInfos(
+      ethCallProvider,
+      chainId,
+      Array.from(targetTokenAddresses),
+    )
+    if (!tokenInfos) return
```

Comment:
Should we add a task to rewrite these fetcher functions to return errors if they're unsuccessful? Currently we return null in a lot of places, but the loadable / useQuery pattern gives us the ability to better handle and surface errors

Reply:
Yeah these queries can be improved. This is a bare minimum port example.